src/mpyl/project_execution.py

```diff
@@ -1,21 +1,33 @@
 """This module contains the ProjectExecution class."""
-import uuid
+
 from dataclasses import dataclass
+from typing import Optional
 
 from .project import Project
 
 
 @dataclass(frozen=True)
 class ProjectExecution:
     project: Project
-    cache_key: str
+    hashed_changes: Optional[str]
     cached: bool
 
     @staticmethod
-    def always_run(project: Project):
-        """Create a ProjectExecution for a project that will always run, regardless of caching"""
+    def create(project: Project, cached: bool, hashed_changes: Optional[str] = None):
+        if cached:
+            return ProjectExecution.skip(project, hashed_changes)
+        return ProjectExecution.run(project, hashed_changes)
+
+    @staticmethod
+    def skip(project: Project, hashed_changes: Optional[str] = None):
+        return ProjectExecution(
+            project=project, hashed_changes=hashed_changes, cached=True
+        )
+
+    @staticmethod
+    def run(project: Project, hashed_changes: Optional[str] = None):
         return ProjectExecution(
-            project=project, cache_key=uuid.uuid4().hex, cached=False
```

Comment:
We no longer set the hashed contents to a random UUID because that makes it impossible to detect after the fact whether there were changes in the project or not.